Merge `evan-goode/bootc` branch to master

evan-goode opened the pull request:
Continued from https://github.com/rpm-software-management/dnf/pull/2180, had to change source branch to avoid rewriting history on the `bootc` branch directly.

<hr>

This PR, along with https://github.com/rpm-software-management/libdnf/pull/1683, merges support for the `--transient` flag and the `persistence=auto|persist|transient` configuration option for https://issues.redhat.com/browse/RHEL-70917. See also https://github.com/rpm-software-management/dnf/pull/2155.

Also included are some build files for Copr that may be useful for future feature branches similar to `bootc`. 

Merging this PR doesn't mean that work on the `--transient` feature is 100% done, just that we're confident enough we're not going to totally scrap it and we want to continue its development on the `master` branch.

evan-goode commented:
This PR requires https://github.com/rpm-software-management/libdnf/pull/1692.

ppisar commented:
When testing the previous pull request, I noticed few strange points about the user interface:

The error message changed from "Error: system is configured to be read-only." to "This bootc system is configured to be read-only." I get that the "error" word was scary. But do we need to talk specifically about bootc? The DNF implementation still uses OSTree API, and commands and I think Fedora's CoreOS is still based on plain rpm-ostreee. But maybe we already discussed this and I forgot the reason why bootc name is good enough.

The error message explains "Pass --transient to perform this and subsequent transactions in a transient overlay which will reset when the system reboots." I'm not a native English, but my understanding is that I call "dnf --transient ..." and then I can call "dnf" without that option until a reboot. That's not how it behaves. It's necessary to pass the --transient option again and again because the overlay remains read-only out DNF file system name space. I find it confusing. Would it be better to change the message to "Pass --transient to perform this transactions in a transient overlay which will reset when the system reboots."? I.e. not mentioning the "subsequent transactions"?

evan-goode commented:
> When testing the previous pull request, I noticed few strange points about the user interface:
> 
> The error message changed from "Error: system is configured to be read-only." to "This bootc system is configured to be read-only." I get that the "error" word was scary. But do we need to talk specifically about bootc? The DNF implementation still uses OSTree API, and commands and I think Fedora's CoreOS is still based on plain rpm-ostreee. But maybe we already discussed this and I forgot the reason why bootc name is good enough.

Yes, the "bootc" name was decided long ago. I think the main reason was to unify the language across RHEL Image Mode. Personally I think you're right that it would be better to use "OSTree" in the language here, but in the future it's conceivable we might add functionality or additional checks here that are bootc-specific. Or for example call `bootc usr-overlay` rather than `ostree admin unlock`, which would introduce a requirement on bootc. None of the new CLI options contain the word "bootc" (`--transient`, `persistence=transient|persist|auto`), it only appears in docs/error messages/status messages, so we could also update it later to "OSTree" if desired.

> The error message explains "Pass --transient to perform this and subsequent transactions in a transient overlay which will reset when the system reboots." I'm not a native English, but my understanding is that I call "dnf --transient ..." and then I can call "dnf" without that option until a reboot. That's not how it behaves. It's necessary to pass the --transient option again and again because the overlay remains read-only out DNF file system name space. I find it confusing. Would it be better to change the message to "Pass --transient to perform this transactions in a transient overlay which will reset when the system reboots."? I.e. not mentioning the "subsequent transactions"?

You're absolutely right, that language was from an earlier version which did affect subsequent transactions. I've changed it per your suggestion (and rewrote history).

jrohel commented on dnf/cli/cli.py:
```diff
@@ -205,28 +205,50 @@ def do_transaction(self, display=()):
+                                      "perform this and subsequent transactions in a transient overlay which "
+                                      "will reset when the system reboots."))
+                        raise CliError(_("Operation aborted."))
+                    assert self.conf.persistence == "transient"
```
There will always be a "transient". A few lines above is `assert self.conf.persistence in ("auto", "transient")`. And "auto" is handled.

 But, OK. The extra `assert` doesn't break anything.

ppisar reviewed (approved):
Thanks for the consolidated patch set. Besides the updated hawkey_version version and the the error message it's identical to the #2180.

The CI failure is caused by the updated hawkey_version value. It is already available since <https://github.com/rpm-software-management/libdnf/pull/1692> but it will be available to CI tomorrow when libdnf will have been rebuilt in nightly.

ppisar commented:
Right now I'm testing this code in a bootc system.

ppisar commented:
I successfully finished the testing.
